components/kyma-environment-broker/common/orchestration/strategies/parallel.go

```diff
@@ -43,80 +46,146 @@ func (p *ParallelOrchestrationStrategy) Execute(operations []orchestration.Runti
 	if len(operations) == 0 {
 		return "", nil
 	}
-	ops := make(chan orchestration.RuntimeOperation, len(operations))
+
 	execID := uuid.New().String()
 	p.mux.Lock()
 	defer p.mux.Unlock()
+	p.scheduleNum[execID] = len(operations)
 	p.wg[execID] = &sync.WaitGroup{}
 	p.dq[execID] = workqueue.NewDelayingQueue()
-
-	if strategySpec.Schedule == orchestration.MaintenanceWindow {
-		sort.Slice(operations, func(i, j int) bool {
-			return operations[i].MaintenanceWindowBegin.Before(operations[j].MaintenanceWindowBegin)
-		})
-	}
-
-	// Send operations to workers
-	for _, op := range operations {
-		ops <- op
+	p.pq[execID] = workqueue.NewDelayingQueue()
+
+	for i, op := range operations {
+		duration, err := p.updateMaintenanceWindow(execID, &operations[i], strategySpec)
+		if err != nil {
+			//error when read from storage or update to storage during maintenance window reschedule
+			p.handleRescheduleErrorOperation(execID, &operations[i])
+			p.log.Errorf("while processing operation %s: %v, will reschedule it", op.ID, err)
+		} else {
+			p.dq[execID].AddAfter(&operations[i], duration)
+		}
 	}
 
 	// Create workers
 	for i := 0; i < strategySpec.Parallel.Workers; i++ {
-		p.createWorker(execID, ops, strategySpec)
+		p.createWorker(execID, strategySpec)
 	}
 
 	return execID, nil
 }
 
-func (p *ParallelOrchestrationStrategy) Wait(executionID string) {
+func (p *ParallelOrchestrationStrategy) createWorker(execID string, strategy orchestration.StrategySpec) {
+	p.wg[execID].Add(1)
+
+	go func() {
+		p.scheduleOperationsLoop(execID, strategy)
+
+		p.mux.RLock()
+		p.wg[execID].Done()
+		p.mux.RUnlock()
+	}()
+}
+
+func (p *ParallelOrchestrationStrategy) scheduleOperationsLoop(execID string, strategy orchestration.StrategySpec) {
 	p.mux.RLock()
-	wg := p.wg[executionID]
+	dq := p.dq[execID]
+	pq := p.pq[execID]
 	p.mux.RUnlock()
-	if wg != nil {
-		wg.Wait()
-	}
-}
 
-func (p *ParallelOrchestrationStrategy) Cancel(executionID string) {
-	if executionID == "" {
-		return
-	}
-	p.log.Infof("Cancelling strategy execution %s", executionID)
-	p.mux.Lock()
-	defer p.mux.Unlock()
-	dq := p.dq[executionID]
-	if dq != nil {
-		dq.ShutDown()
+	for {
+		p.mux.RLock()
+		if p.scheduleNum[execID] <= 0 {
+			p.mux.RUnlock()
+			dq.ShutDown()
+			pq.ShutDown()
+			break
+		}
+		p.mux.RUnlock()
+
+		if dq.ShuttingDown() || pq.ShuttingDown() {
```

Comment:
I think this is not needed because of the shutdown check on dq.Get() below. Once dq is marked for shutdown,. it will always instantly report shutdown=true for all calls to Get() once the dq is drained.
And dq has to be drained correctly in the case of orchestration cancel (i.e. calling `ParallelOrchestrationStrategy.Cancel()`), so we must not break prematurely.

Reply:
yes.. this is not needed. also thought about it but still keep it.. yes, Get() will also return shutdown as true and break.. will remove it.

Reply:
removed this condition and together with the break when `p.scheduleNum[execID] <= 0`